File: leetcode/longest_valid_parentheses/LongestValidParentheses.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        int last = -1; 
        int maxLength = 0; 
        stack<int> stk; 
        
        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '(') {
                stk.push(i); 
            }
            else {
                if (stk.size() == 0) {  // ) is a seperator 
                    last = i; 
                }
                else {  // Match 
                    stk.pop(); 
                    
                    if (stk.size() == 0) {  // After matching, stack becomes empty
                        maxLength = max(maxLength, i - last); 
                    }
                    else {
                        maxLength = max(maxLength, i - stk.top()); 
                    }
                }
            }
        }
        
        return maxLength; 
    }
};

File: leetcode4/LongestValidParentheses.java
public class Solution {
    public int longestValidParentheses(String s) {
        int max = 0;
        Stack<Integer> stack = new Stack<Integer>();
        int len = s.length();
        int last = -1;

        for (int i = 0; i < len; i++) {
            if (s.charAt(i) == '(') {
                stack.push(i);
            }
            else {
                if (stack.size() != 0) {
                    stack.pop();

                    if (stack.size() != 0) {
                        max = Math.max(max, i - stack.peek());
                    } else {
                        max = Math.max(max, i - last);
                    }
                } else {
                    last = i;
                }
            }
        }

        return max;
    }
}
